### DISCUSSION TITLE
How to get GPU passthrough working on WSL2 with Windows 11 as hosts , Ubuntu 22.10 as guest os and FreeBSD virtualized with qemu-kvm

### DISCUSSION THREAD

**Discussion by Marietto2008** [microsoft/WSL#10334] [General]
### Windows Version

Windows 11 PRO Version 10.0.22000.2176

### WSL Version

Versione WSL: 1.3.11.0

### Are you using WSL 1 or WSL 2?

- [X] WSL 2
- [ ] WSL 1

### Kernel Version

5.15.90.2-3

### Distro Version

Ubuntu 22.10

### Repro Steps

After a week of hard work,I've just configured my WSL2 subsystem as I want : I have enabled the ZFS support,I can start the graphical applications directly from the Ubuntu command line ; I've installed the nvidia driver and the CUDA libraries within Ubuntu and finally I've configured qemu/kvm and virt-manager so that I can virtualize FreeBSD on top of Ubuntu. Now,the problem is : how can I "tell" to FreeBSD to run Xorg with one of the GPUs that I have on the PC. Actually I have two GPUs on the system. The GTX 1060,that I use as default,where I have attached my monitor and the RTX 2080 ti,that usually is used to be passed into another OS. When I'm using Linux as host os,I would like to virtualize Windows and I passthru the RTX 2080 ti in Windows using qemu-kvm and virt-manager ; when I use FreeBSD as host os,I want to pass thru the RTX 2080 ti to a Linux guest os using bhyve. But when I use Windows 11 as host os and Ubuntu as guest,how can I pass thru my RTX 2080 ti with qemu-kvm and virt-manager to FreeBSD for example ? Below you can see all the PCI devices that are recognized within Ubuntu 22.04 at the moment :
```

# lspci
0301:00:00.0 SCSI storage controller: Red Hat, Inc. Virtio filesystem (rev 01)
05c0:00:00.0 SCSI storage controller: Red Hat, Inc. Virtio filesystem (rev 01)
117a:00:00.0 SCSI storage controller: Red Hat, Inc. Virtio filesystem (rev 01)
242b:00:00.0 3D controller: Microsoft Corporation Basic Render Driver
4e53:00:00.0 SCSI storage controller: Red Hat, Inc. Virtio filesystem (rev 01)
5dc2:00:00.0 SCSI storage controller: Red Hat, Inc. Virtio filesystem (rev 01)
6164:00:00.0 SCSI storage controller: Red Hat, Inc. Virtio filesystem (rev 01)
6251:00:00.0 SCSI storage controller: Red Hat, Inc. Virtio filesystem (rev 01)
7c22:00:00.0 SCSI storage controller: Red Hat, Inc. Virtio filesystem (rev 01)
840d:00:00.0 SCSI storage controller: Red Hat, Inc. Virtio filesystem (rev 01)
94be:00:00.0 SCSI storage controller: Red Hat, Inc. Virtio filesystem (rev 01)
ad48:00:00.0 SCSI storage controller: Red Hat, Inc. Virtio filesystem (rev 01)
ae16:00:00.0 SCSI storage controller: Red Hat, Inc. Virtio filesystem (rev 01)
af96:00:00.0 SCSI storage controller: Red Hat, Inc. Virtio filesystem (rev 01)
baff:00:00.0 SCSI storage controller: Red Hat, Inc. Virtio filesystem (rev 01)
d5cc:00:00.0 SCSI storage controller: Red Hat, Inc. Virtio filesystem (rev 01)
d9a5:00:00.0 System peripheral: Red Hat, Inc. Virtio file system (rev 01)
de65:00:00.0 SCSI storage controller: Red Hat, Inc. Virtio console (rev 01)
fd8d:00:00.0 3D controller: Microsoft Corporation Basic Render Driver
```

the only two virtual graphics adapter that can be used are the following :

```
 242b:00:00.0 3D controller: Microsoft Corporation Basic Render Driver
 fd8d:00:00.0 3D controller: Microsoft Corporation Basic Render Driver
```

I've already tried to configure Xorg in FreeBSD declaring inside the xorg.conf the one and only graphic "card" I see,the "00:01.0 VGA Compatible Controller RedHat QXL paravirtual graphic card" :

![1](https://github.com/microsoft/WSL/assets/755339/ae8a877f-1ffa-4738-94e5-fd805089761e)

like this :

![2](https://github.com/microsoft/WSL/assets/755339/3233581a-c636-4ef8-a8a4-857a7ae099ff)

unfortunately it is not accepted as a valid graphic card :

![3](https://github.com/microsoft/WSL/assets/755339/8d4a198d-9e45-481b-886f-075e20b44041)

What shoud I do to have a working graphic enviroment running on FreeBSD ? Maybe should I try to pass-thru one of these two graphic adapters

```
242b:00:00.0 3D controller: Microsoft Corporation Basic Render Driver
fd8d:00:00.0 3D controller: Microsoft Corporation Basic Render Driver
```

in virt-manager like they were a physical graphic cards ? I'm very doubtful that it works like this.

Anyway,these are the physical GPUs that are on my system :

<img width="769" alt="4" src="https://github.com/microsoft/WSL/assets/755339/5086cb2d-8b87-48fe-a827-40cb07846c02">

both are recognized on Windows 11 and on Ubuntu :

```
# nvidia-smi

+--------------------------------------------------------------------------------------------+
| NVIDIA-SMI 536.40                 Driver Version: 536.40       CUDA Version: 12.2               
|-----------------------------------------+----------------------+---------------------------+
| GPU  Name                     TCC/WDDM  | Bus-Id        Disp.A | Volatile Uncorr. ECC          
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M.  
|                                         |                      |               MIG M.                                              
|=========================================+=========================================
|   0  NVIDIA GeForce GTX 1060 3GB  WDDM  | 00000000:01:00.0  On |                  N/A  
| 57%   41C    P2              25W / 120W |    710MiB /  3072MiB |      0%      Default            
|                                         |                      |                  N/A                                                  
+-----------------------------------------+----------------------+---------------------------+
|   1  NVIDIA GeForce RTX 2080 Ti   WDDM  | 00000000:02:00.0 Off |                  N/A      
| 31%   40C    P8              21W / 250W |     74MiB / 11264MiB |      0%                              
+_________________________________________+______________________+____________________________
```

Some background :

```

    Windows Version = Windows 11 PRO Version 10.0.22000.2176

    WSL2 Version = 1.3.11.0

    Kernel Version = 5.15.90.2-3-microsoft-standard

    Distro Version = Ubuntu 22.10

    Version kernel: 5.15.90.2-3

    Version WSLg: 1.0.54

    Version MSRDC: 1.2.4240

    Version Direct3D: 1.608.2-61064218

    Version DXCore: 10.0.25880.1000-230602-1350.main

```

### Expected Behavior

That I can use the RTX 2080 ti within FreeBSD os as default gpu.

### Actual Behavior

Unfortunately my gpus aren't exposed on the WSL2 sub system.

**Comment by moekhalil** [comment]
IIRC, WSL2 uses GPU partitioning (GPU-PV) instead of PCIe passthrough. 

nvidia-SMI works and CUDA works via that GPU "Wrapper", but it's not actual passthrough.

If the host is required to be Windows, I believe Hyper-V (and maybe -- depending on needs -- Windows Server) might be needed here.

**Comment by Marietto2008** [comment]
Hello.

I'm trying to use Ubuntu VM on a Hyper-V with Microsoft GPU-P support. The tutorial that I have used to reach the goal is the one below :

https://github.com/brokeDude2901/dxgkrnl_ubuntu

The result that I have achieved has been to enable the nVidia driver and CUDA libraries within an Ubuntu 20.04 VM,but Blender Cycles does not recognize my GPUs :

![2023-07-20 22_39_34-NVIDIA GeForce Overlay](https://github.com/microsoft/WSL/assets/755339/79fdd711-3c43-4215-87b8-17815becd30a)

Do you know the reason ? Instead,Blender Cycles recognizes at least one GPU (I have 2 GPUs,but it recognizes only one : but I have an idea about the reason : on the script I have declared 16 GB of memory instead of the default,32 GB) if the VM is Windows 11 :

![2023-07-20 23_28_55-Windows-11-GNUP on MARIETTO-WINDOW - Virtual Machine Connection](https://github.com/microsoft/WSL/assets/755339/c2de6a7a-7fe3-45fa-a3c6-4facb85a172c)

Something is missing or it does not work well if the VM is Ubuntu 20.04. I know what it is. While Windows 11 recognizes the GPU directly on the Device Manager,Ubuntu does not recognise it on the terminal when I issue the command "lspci" as you can see below :

```
root@marietto-HyperV:/home/marietto# lspci

7ea5:00:00.0 3D controller: Microsoft Corporation Device 008e
ec5c:00:00.0 3D controller: Microsoft Corporation Device 008e
```

what I can do to allow Blender Cycles to recognize my GPU as a good GPU for rendering ?

**Comment by moekhalil** [comment]
@Marietto2008  -- Have you tried using Ubuntu under WSL2 with Cuda installed?

Just installed Blender on my WSL2 Ubuntu distro, it is showing CUDA GPU just fine. 
Windows 11 23593 + WSL2 + Ubuntu 23.04

<img width="955" alt="image" src="https://github.com/microsoft/WSL/assets/5835530/d0dae60f-0b8b-428f-879b-60d10de256e4">

**Comment by Marietto2008** [comment]
yes,on the WSL2 my GPUs are recognized by Blender Cycles :

![2023-07-21 22 07 15](https://github.com/microsoft/WSL/assets/755339/2c19f970-a6ff-4a26-b8cf-714d9209dfa0)

but look at the bottom of the photo : OpenGL renderer string should shows the name of my GPU,but it isn't there.

**Comment by Marietto2008** [comment]
I don't understand. Why my GPUs are recognized within Ubuntu by BLender cycles but it is not inside Ubuntu installed with Hyper-V ?They don't use the same technology ?

**Comment by moekhalil** [comment]
No. WSL2 does not support hardware pass-thru like Hyper-V. 
The Nvidia + Cuda drivers on WSL2 are just lightweight wrappers/linkers that send off GPU instructions to the windows graphics driver through symlinking: `/lib/wsl/lib/libcuda or libnvidia` to `C:\Windows\System32\lxss\lib\libcuda or libnvidia`

**Comment by Marietto2008** [comment]
ok,but my motherboard does not support SR-IOV,I can't pass thru my graphic card using the Discrete Device Assignement. For this reason I'm trying to share the GPU between the host os (WIndows 11) and the guest os (in this case Ubuntu,because if I use Windows 11 as guest os,it works). This is the tutorial that allows to do that :

https://github.com/brokeDude2901/dxgkrnl_ubuntu

unfortunately Blender Cycles can't render anything because it does not recognize the GPUs,even if the cuda libraries works os,they are able to recognize both my GPUs. I don't know if it is natural of I'm making some mistake. This bring us to one of my old comment :

https://github.com/microsoft/WSL/issues/10294#issuecomment-1644644218

**Comment by Marietto2008** [comment]
`# glxinfo | grep "OpenGL renderer string" display D3D12 (My GPU Name) :`

![2023-07-22 15_53_40-Ubuntu-2104 on MARIETTO-WINDOW - Virtual Machine Connection](https://github.com/microsoft/WSL/assets/755339/855bd4dd-c9e4-44f3-9333-299d60239f43)

**Comment by fedecompa** [comment]
I have a similar problem. I installed all the packages from Intel OneAPI link. On Windows 11 all works fine and the VideoProcessor Intel(R) UHD Graphics can be used from OpenVINO libraries. On the contrary on WSL2 with the Ubuntu 22.04 distro I see no GPU:

  clinfo
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_GPU)  No devices found in platform
  
   sycl-ls
   [opencl:cpu][opencl:0] Intel(R) OpenCL, Intel(R) Core(TM) i7-10850H CPU @ 2.70GHz OpenCL 3.0 (Build 0) [2024.18.7.0.11_160000]
   
   
   https://www.intel.com/content/www/us/en/docs/oneapi/installation-guide-linux/2023-0/configure-wsl-2-for-gpu-workflows.html

**Comment by elsaco** [comment]
@fedecompa before running `clinfo` make sure you're using the correct driver. Running `glxinfo -B` inside WSL should display `Device: D3D12` otherwise if software rendering is used, it won't work.

Sample output on my system:

```
elsaco@eleven:~$ glxinfo -B
name of display: :0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Microsoft Corporation (0xffffffff)
    Device: D3D12 (Intel(R) UHD Graphics 750) (0xffffffff)
    Version: 23.2.1
    Accelerated: yes
---cut---
```
The main dGPU in my system is a RTX 4070 but the graphics inside WSL is provided by the iGPU of the processor. Notice the device description as 'Intel(R) UHD Graphics 750`

Some `clinfo` ouput:

```
elsaco@eleven:~$ clinfo -l
Platform #0: Intel(R) OpenCL Graphics
 `-- Device #0: Intel(R) Graphics [0x4c8a]
```

```
elsaco@eleven:~$ clinfo
Number of platforms                               1
  Platform Name                                   Intel(R) OpenCL Graphics
  Platform Vendor                                 Intel(R) Corporation
  Platform Version                                OpenCL 3.0
  Platform Profile                                FULL_PROFILE
---cut---
```

The kernel driver for the virtual 3D controller inside WSL should be `dxgkrnl`:

```
elsaco@eleven:~$ lspci -k
3aa9:00:00.0 3D controller: Microsoft Corporation Device 008e
        Kernel driver in use: dxgkrnl
4c5c:00:00.0 System peripheral: Red Hat, Inc. Virtio file system (rev 01)
        Subsystem: Red Hat, Inc. Virtio file system
        Kernel driver in use: virtio-pci
8ce4:00:00.0 SCSI storage controller: Red Hat, Inc. Virtio console (rev 01)
        Subsystem: Red Hat, Inc. Virtio console
        Kernel driver in use: virtio-pci
efb6:00:00.0 3D controller: Microsoft Corporation Device 008e
        Kernel driver in use: dxgkrnl

```

**Reply by fedecompa** [reply]
@elsaco  this is the output of the command  glxinfo -B:

name of display: :0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Mesa (0xffffffff)
    Device: llvmpipe (LLVM 15.0.7, 256 bits) (0xffffffff)
    Version: 23.2.1
    Accelerated: no
    Video memory: 7945MB
    Unified memory: yes
    Preferred profile: core (0x1)
    Max core profile version: 4.5
    Max compat profile version: 4.5
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.2
Memory info (GL_ATI_meminfo):
    VBO free memory - total: 31 MB, largest block: 31 MB
    VBO free aux. memory - total: 6629 MB, largest block: 6629 MB
    Texture free memory - total: 31 MB, largest block: 31 MB
    Texture free aux. memory - total: 6629 MB, largest block: 6629 MB
    Renderbuffer free memory - total: 31 MB, largest block: 31 MB
    Renderbuffer free aux. memory - total: 6629 MB, largest block: 6629 MB
Memory info (GL_NVX_gpu_memory_info):
    Dedicated video memory: 4294448247 MB
    Total available memory: 4294456192 MB
    Currently available dedicated video memory: 31 MB
OpenGL vendor string: Mesa
OpenGL renderer string: llvmpipe (LLVM 15.0.7, 256 bits)
OpenGL core profile version string: 4.5 (Core Profile) Mesa 23.2.1-1ubuntu3.1~22.04.2
OpenGL core profile shading language version string: 4.50
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile

OpenGL version string: 4.5 (Compatibility Profile) Mesa 23.2.1-1ubuntu3.1~22.04.2
OpenGL shading language version string: 4.50
OpenGL context flags: (none)
OpenGL profile mask: compatibility profile

OpenGL ES profile version string: OpenGL ES 3.2 Mesa 23.2.1-1ubuntu3.1~22.04.2
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.20

___________________________________________________________________________________


This is the output of the command  clinfo -l: 

Platform #0: Intel(R) OpenCL
 `-- Device #0: Intel(R) Core(TM) i7-10850H CPU @ 2.70GHz
 
 __________________________________________________________________________________
 
 This is the output of the command  lspci -k:
 
1ede:00:00.0 SCSI storage controller: Red Hat, Inc. Virtio 1.0 console (rev 01)
        Subsystem: Red Hat, Inc. Virtio console
        Kernel driver in use: virtio-pci
2d05:00:00.0 3D controller: Microsoft Corporation Basic Render Driver
        Kernel driver in use: dxgkrnl
b034:00:00.0 System peripheral: Red Hat, Inc. Virtio file system (rev 01)
        Subsystem: Red Hat, Inc. Virtio file system
        Kernel driver in use: virtio-pci
        
_________________________________________________________________________________

Which drivers should I install in the distro  Ubuntu 22.04.5 LTS (GNU/Linux 5.15.153.1-microsoft-standard-WSL2 x86_64) to let it see the Intel(R) UHD Graphics device? On windows 11 I have the driver version 27.20.100.9664
Thanks.

**Comment by fedecompa** [comment]
I solved updating the Intel graphics drivers to latest version for WSL2:

https://docs.openvino.ai/archive/2023.1/openvino_docs_install_guides_configurations_for_intel_gpu.html#additional-resources
